### ISSUE TITLE
Sending two save commands in parallel to a vw daemon results in no model

### ISSUE THREAD

**Issue by joachimdb** [JohnLangford/vowpal_wabbit#627]
Trace of vw interaction:

Joachims-MacBook-Air:vw-socket-issue joachim$ vw -f model.vw --save_resume --daemon --loss_function logistic --learning_rate 5.0 --decay_learning_rate 0.95 --power_t 0.5 --initial_t 0 --l1 1e-8 --l2 1e-8 
using l1 regularization = 1e-08
using l2 regularization = 1e-08
final_regressor = model.vw
Num weight bits = 18
learning rate = 5
initial_t = 0
power_t = 0.5
using no cache
calling accept

Joachims-MacBook-Air:vw-socket-issue joachim$ nc -c localhost 26542 < resources/train-examples.txt > /tmp/out
Joachims-MacBook-Air:vw-socket-issue joachim$ echo "save" | nc -c localhost 26542 && sleep 5 && ls -l model.vw
reading data from port 26542
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
saving regressor to model.vw

-rw-r--r--  1 joachim  staff  328970 May 12 14:53 model.vw

Joachims-MacBook-Air:vw-socket-issue joachim$ echo "save" | nc -c localhost 26542 && echo "save" | nc -c localhost 26542 && sleep 10 && ls -l model.vw
ls: model.vw: No such file or directory

**Comment by joachimdb** [comment]
Update: this issue cannot be reproduced on Ubuntu, see https://github.com/mmroden/vw_daemon_tests/blob/master/features/parallel_saves.feature

**Comment by JohnLangford** [comment]
This seems like a "don't do that".

VW creates a temporary file, saves the contents into that file, then 
moves it to the final file.  This process ensures integrity when there 
is an incomplete write and happens here: 
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/parse_regressor.cc#L239

What happens to the OS when it's told to open-with-create the same file 
twice? It could be OS-dependent.

-John

**Comment by mmroden** [comment]
Maybe there could be a lock on saving, so it can only happen one at a time?

**Comment by JohnLangford** [comment]
That would be fine---Want to implement it?

It seems you need to either have an external or an internal lock. In 
many circumstances, an external lock seems easy.

-John

**Comment by JohnLangford** [comment]
Closing this for now.  I would welcome a pull request, but am too overloaded to handle myself at the moment.
